file/sequence_file_reader.cc

```diff
@@ -25,19 +26,32 @@ namespace ROCKSDB_NAMESPACE {
 IOStatus SequentialFileReader::Create(
     const std::shared_ptr<FileSystem>& fs, const std::string& fname,
     const FileOptions& file_opts, std::unique_ptr<SequentialFileReader>* reader,
-    IODebugContext* dbg) {
+    IODebugContext* dbg, RateLimiter* rate_limiter) {
   std::unique_ptr<FSSequentialFile> file;
   IOStatus io_s = fs->NewSequentialFile(fname, file_opts, &file, dbg);
   if (io_s.ok()) {
-    reader->reset(new SequentialFileReader(std::move(file), fname));
+    reader->reset(new SequentialFileReader(std::move(file), fname, nullptr, {},
+                                           rate_limiter));
   }
   return io_s;
 }
 
-IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch,
+                                    Env::IOPriority rate_limiter_priority) {
   IOStatus io_s;
+  uint64_t filesize;
+  IOOptions io_opts;
+  // Need to get file size for each read since new content might be added to
+  // file after last read
+  io_s = file_->GetFileSize(filesize);
+  if (!io_s.ok()) {
+    return io_s;
```

Comment:
cc @anand1976, @cbi42 - I wonder if we should fail the read just because an optional API `GetFileSize` is not implemented. Plus it is only needed when there's a need for rate-limiting. Even when we actually need rate-limiting, is this a strong enough reason to fail read?

An alternative is to fall back to rate-limiting not knowing the file size (so no `n = std::min(n, static_cast<size_t>(filesize) - offset_);` below, which leads to some extra unnecessary rate-limit request) but does not fail read.

Reply:
I don't feel "io_s = file_->GetFileSize(filesize);" will slow down `SequentialFileReader::Read` much but I'd still suggest:
(1) Move it into if-branch when rate-limiting is needed as it's only used for rate-limiting purpose so far
(2) db bench or microbench on backup/restore for pre and (post PR + instrumentally mocking the actual wait in rate-limiter (e.g, temporarily remove https://github.com/facebook/rocksdb/blob/7.2.fb/util/rate_limiter.cc#L38 )). The reason to mock out the actual wait in rate-limiter is that we only interested in knowing how much `io_s = file_->GetFileSize(filesize);` will slow down read. 

Although I don't anticipate regression from (2) especially with the optimization you are exploring right now, I'd still stay cautious and run some benchmark.

For more stable db_bench running, you can leverage a recent feaeture like https://github.com/facebook/rocksdb/pull/9882

Reply:
@anand1976 
> No one should be simultaneously reading and writing the file.

I found [here](https://github.com/facebook/rocksdb/blob/c4cd8e1accfd010bba9a919e234ba04505071cc1/db/log_reader.h#L94 
) that log_reader has this `UnmarkEof` function, and it will read from a `SequentialFileReader` if it knows that more data is appended to the file. Do you think that's a legit use case?

Regarding `GetFileSize() need not be called for every read. I'm working to improve this with @hx235's suggestion above and only refresh file size when reading beyond end of file.

> we should call fs->GetFileSize() to get the size rather than add a new API in FSSequentialFile

Good point, thanks, I'll try this change.

Reply:
> I found here that log_reader has this UnmarkEof function, and it will read from a SequentialFileReader if it knows that more data is appended to the file. Do you think that's a legit use case?

That's a good point. Maybe the secondary instance functionality uses it to tail the MANIFEST, among other things. @riversand963 is the expert in that.

Reply:
I suspect we currently do not require `GetFileSize()` even once per file. It might be doing that on local filesystem if you strace a backup, but when run on an Env/FileSystem that implements GetChildrenFileAttributes() it should not. The motivation is, on distributed file systems like HDFS, querying file size goes through the namenode, which can be (and has been at times) the scalability bottleneck.

So I guess the question is, is there a way to provide file size that backup/checkpoint already got from `GetChildrenFileAttributes()` to as the initial value for `SequentialFileReader`'s size?

Reply:
If we can get file size from `GetChildrenFileAttributes()`, this would save the initial call. However, we still need another call when reading beyond the last byte to check if new content was added to the file, so it will still be once per file?

Reply:
Oh ok. Maybe just try to prevent regression in a common existing use case (backup local flash <-> remote FS). In that case the restore operation is the relevant one because in that case we are reading files from remote FS. I think we can prevent reading past EOF by passing `file_info->size` as `size_limit` here https://github.com/facebook/rocksdb/blob/f6339de0d2e0383faa90fdf011a665c94be8f692/utilities/backup/backup_engine.cc#L1925

Reply:
I wasn't following what test failures this GetFileSize() intends to address. Have you considered the alternative of restricting how much bytes clients of SequentialFileReader call for Read()? Looks like CopyFile() already does this (based on its own GetFileSize()), and BackupEngine could too as it also knows file size. Are there more places similar logic would be needed to fix the test failures? Or, are there other places we are worried about over-charging?

Reply:
@ajkr I have not tried the alternative, and I think limiting the size should work for CopyFile() and BackupEngine. The worry is for over-charging in log reader, as it reads a log file that may be written to at the same time. Maybe that's fine (currently I used IO_TOTAL for it), and I can add a TODO for fixing that overcharge later.

Reply:
To summarize, I have updated the PR to restrict BackupEngine (the main user of SequentialFileReader) on how many bytes it calls Read() for (so no additional system call is introduced). Specifically, I updated the `size_limit` parameter [here](https://github.com/facebook/rocksdb/blob/987ca52c9985afb3c085ed5cd272921dfc1165aa/utilities/backup/backup_engine.cc#L1436) and [here](https://github.com/facebook/rocksdb/blob/987ca52c9985afb3c085ed5cd272921dfc1165aa/utilities/backup/backup_engine.cc#L1861). 
I have added TODOs in `log_reader` about that it might overcharge `SequentialFileReader`'s rate limiter and will leave it for future work (it currently uses `IO_TOTAL` priority so this PR should not affect it). 
Thanks for the review and suggestions!